File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        # Idea: For each bar, we store the left and right boundary to calculate the area made using that bar
        n = len(heights)
        stack = []
        res = 0

        for i, h in enumerate(heights):
            start = i
            # Strictly increasing stack; the new bar can take the old index to store the left boundary
            while stack and h <= stack[-1][1]:
                idx, height = stack.pop()
                res = max(res, height * (i - idx))
                start = idx
            stack.append([start, h])

        # The right boundary is now the end of the histogram
        for i, h in stack:
            res = max(res, h * (n - i))
        
        return res

File: Data Structures & Algorithms/trapping-rain-water/submission-1.js
class Solution {
    /**
     * @param {number[]} height
     * @return {number}
     */
    trap(height) {
        let res = 0;
        let start = 0, end = height.length - 1;
        let leftMax = height[0], rightMax = height[height.length - 1];

        while (start < end) {
            if (leftMax < rightMax) {
                start += 1;
                leftMax = Math.max(leftMax, height[start]);
                res += leftMax - height[start];
            } else {
                end -= 1;
                rightMax = Math.max(rightMax, height[end]);
                res += rightMax - height[end];
            }
        }

        return res;
    }
}
